dashboard/legacy/middleware/files_api.rb

```diff
@@ -424,8 +426,16 @@ def put_file(endpoint, encrypted_channel_id, filename, body, project_type = nil)
     if (endpoint == 'libraries' && file_type != '.java') || profanity_project_type?(project_type)
       begin
         if profanity_project_type?(project_type)
+          # For aichat lab, we only check the student system prompt and retrieval contexts.
           text_to_check = get_text_for_profanity_check(project_type, body)
-          share_failure = ShareFiltering.find_profanity_failure(text_to_check, request.locale)
+          locale_code = request.locale.to_s.split('-').first
+          # Moderate user text for toxicity.
+          # get_toxicity returns an object with the following fields:
+          # text: string, toxicity: number, and max_category {name: string, score: number}
+          text_to_check_comprehend_response = AichatComprehendHelper.get_toxicity(text_to_check, locale_code)
+          if text_to_check_comprehend_response[:toxicity] >= get_toxicity_threshold_user_sources
+            share_failure = ShareFailure.new(ShareFiltering::FailureType::PROFANITY, text_to_check_comprehend_response)
+          end
```

Comment:
Could we move the aichat/comprehend-specific code into a separate function, and/or maybe combine it with `get_text_for_profanity_check`? For example, maybe a function like

```
# Check for profanity in project sources if needed
def find_project_profanity(locale)
  # Currently, only AI Chat is checked for profanity
  if project_type == 'aichat'
    source = JSON.parse(body)['source']
    source_json = JSON.parse(source)
    text = source_json['systemPrompt'] + ' ' + source_json['retrievalContexts'].join(' ')
    # Use Amazon Comprehend to check AI Chat contents for toxicity
    comprehend_response = AichatComprehendHelper.get_toxicity(text, locale)
    if comprehend_response[:toxicity] = DCDO.get('aichat-toxicity-threshold-user-sources')
      return ShareFailure.new(ShareFiltering::FailureType::PROFANITY, comprehend_response)
    end
  # else
  #   in theory, if/when we want to check other labs with WebPurify, that can go here 
  end
end
```

And then we just call `share_failure = find_project_profanity` here

Reply:
I like this idea - updated. Thanks!